File: java/Pow.java
/*
 * Filename: Pow.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 13 May 2015 01:45:24 PM EDT
 */

public class Solution {
	public double myPow(double x, int n) {
		if (n == 0) {
			return 1;
		}

		double val = myPow(x, n / 2);

		if (n < 0) {
			x = 1 / x;
			n = -n;
		}

		if (n % 2 == 0) {
			return val * val;
		} else {
			return val * val * x;
		}
	}
}

File: cpp/pow.cpp
/*
 * Filename: pow.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 26 Jan 2016 10:13:15 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution {
	public:
		double myPow(double x, int n) {
			if (n == 0)
			{
				return 1;
			}

			double val = myPow(x, n / 2);
			if (n < 0)
			{
				x = 1 / x;
				n = -n;
			}

			if (n % 2)
			{
				return val * val * x;
			}
			else
			{
				return val * val;
			}
		}
};
